File: 744.find-smallest-letter-greater-than-target.java
/*
 * @lc app=leetcode id=744 lang=java
 *
 * [744] Find Smallest Letter Greater Than Target
 */

// @lc code=start
class Solution {
    public char nextGreatestLetter(char[] letters, char target) {
        int minDiff = Integer.MAX_VALUE;
        char letter = 0;
        for (char ch : letters) {
            if (ch <= target) continue;
            int diff = ch - target;
            if (diff < minDiff) {
                minDiff = diff; letter = ch;
            }
        }
        return (minDiff == Integer.MAX_VALUE) ? letters[0] : letter;
    }
}
// @lc code=end

File: 744.find-smallest-letter-greater-than-target.cpp
/*
 * @lc app=leetcode id=744 lang=cpp
 *
 * [744] Find Smallest Letter Greater Than Target
 */

// @lc code=start

#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        auto pos = upper_bound(letters.begin(), letters.end(), target);
        return (pos == letters.end()) ? letters.front() : *pos;
    }
};
// @lc code=end
